### PR TITLE
PR: Refactor to separate ShellWidget from VariableExplorerWidget and FigureBrowser

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

`ShellWidget` has explicit references to `VariableExplorerWidget` (`shellwidget.namespacebrowser`) and `FigureBrowser` (`shellwidget.figurebrowser`). This refactoring removes these references to avoid references loops and increase the separation between plugins.

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 1]
```diff
@@ -1018,3 +1016,56 @@ def focusOutEvent(self, event):
         """Reimplement Qt method to send focus change notification"""
         self.sig_focus_changed.emit()
         return super(ShellWidget, self).focusOutEvent(event)
+
+    # --- value access --------------------------------------------------------
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/shell.py:1020]: I suggest that we move back the methods in this section to `NamepaceBrowserWidget`. They are only necessary for the Variable Explorer and not called by other methods here.

As with `DebuggingWidget` and the other widgets that compose `ShellWidget`, that allows us to have the size of `ShellWidget` manageable.

File: spyder/plugins/variableexplorer/widgets/namespacebrowser.py [hunk 2]
```diff
@@ -40,6 +42,8 @@
 
 # Constants
 VALID_VARIABLE_CHARS = r"[^\w+*=¡!¿?'\"#$%&()/<>\-\[\]{}^`´;,|¬]*\w"
+# Max time before giving up when making a blocking call to the kernel
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/namespacebrowser.py:45]:
```suggestion

# Max time before giving up when making a blocking call to the kernel
```

File: spyder/plugins/framesexplorer/widgets/main_widget.py [hunk 3]
```diff
@@ -228,7 +229,9 @@ def create_new_widget(self, shellwidget):
         widget.sig_hide_finder_requested.connect(self.hide_finder)
         widget.sig_update_actions_requested.connect(self.update_actions)
 
-        widget.sig_show_namespace.connect(shellwidget.set_namespace_view)
+        widget.sig_show_namespace.connect(
+            lambda namespace: self.sig_show_namespace.emit(
+                namespace, shellwidget))
```
Comment (ccordoba12) [inline spyder/plugins/framesexplorer/widgets/main_widget.py:234]: When I merged your initial PR for Frames I didn't understand why you added an action to show the namespace of the current frame.

I mean, my question: why can't we do that automatically?

Comment (impact27) [reply spyder/plugins/framesexplorer/widgets/main_widget.py:234]: Let's say I capture the state of the program. I have 3 threads with a few frames. If I want to see the namespace of the last frame of thread number two, I can show it in the variable explorer

File: spyder/plugins/ipythonconsole/widgets/shell.py (hunk 1 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/shell.py]: Done

File: spyder/plugins/plots/widgets/figurebrowser.py [hunk 4]
```diff
@@ -238,7 +241,7 @@ def change_auto_fit_plotting(self, state):
     def set_shellwidget(self, shellwidget):
         """Bind the shellwidget instance to the figure browser"""
         self.shellwidget = shellwidget
-        shellwidget.set_figurebrowser(self)
+        self.shellwidget.mute_inline_plotting = self.mute_inline_plotting
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/plots/widgets/figurebrowser.py]: Done

**Review by ccordoba12 — APPROVED** [review]
Looks good to me now, thanks @impact27!

**Pull request merged**